File: child_android/app/src/main/kotlin/com/guardnest/kid/WebFilter.kt
package com.guardnest.kid

/**
 * Holds the active web-filter policy in memory so the accessibility service can
 * block bad sites straight from the browser address bar — no VPN, so it doesn't
 * slow the internet down.
 *
 * The whole policy lives in one immutable [Policy] behind a single volatile
 * reference. It used to be several independent `@Volatile` fields updated one
 * after another by the Firestore listeners, which meant a check running on the
 * accessibility thread could read a half-applied policy (e.g. new categories
 * with the old enabled flag) and let a blocked page through. Swapping one
 * snapshot makes every read self-consistent.
 *
 * Fed by [EnforcementService]: the per-family custom sites from
 * `rules/webFilter`, and the global switches from `appConfig/webPolicy`.
 */
object WebFilter {

    /**
     * One consistent view of the filter. Replaced wholesale, never mutated.
     *
     * "Only allow one browser" ([blockOtherBrowsers]): when on, every browser
     * except [approvedBrowser] is blocked, which closes the private/incognito
     * gap for all browsers. Safe browsing implies this with Chrome as the
     * approved browser, so the filter (which works via Chrome's address bar)
     * can't be bypassed by using a different browser or its private mode.
     */
    data class Policy(
        val enabled: Boolean = false,
        val blockedSites: Set<String> = emptySet(), // exact domains from the parent
        val enabledCategories: Set<String> = emptySet(),
        val blockOtherBrowsers: Boolean = false,
        val approvedBrowser: String? = null,
    )

    @Volatile
    private var policy = Policy()

    /** The current snapshot. Read once per decision so it can't change midway. */
    fun policy(): Policy = policy

    val enabled: Boolean get() = policy.enabled
    val blockedSites: Set<String> get() = policy.blockedSites
    val enabledCategories: Set<String> get() = policy.enabledCategories
    val blockOtherBrowsers: Boolean get() = policy.blockOtherBrowsers

    /** Applies the per-family custom blocked sites (from `rules/webFilter`). */
    @Synchronized
    fun updateSites(sites: Set<String>) {
        policy = policy.copy(blockedSites = sites)
    }

    /**
     * Applies the site admin's global policy (from `appConfig/webPolicy`) as one
     * atomic change, so the safe-browsing master, the browser lock and the
     * blocked categories can never be observed out of step with each other.
     */
    @Synchronized
    fun updatePolicy(
        enabled: Boolean,
        blockOtherBrowsers: Boolean,
        categories: Set<String>,
        approvedBrowser: String? = policy.approvedBrowser,
    ) {
        policy = policy.copy(
            enabled = enabled,
            blockOtherBrowsers = blockOtherBrowsers,
            enabledCategories = categories,
            approvedBrowser = approvedBrowser,
        )
    }

    /** The browser package the child is allowed to use (defaults to Chrome). */
    fun effectiveApprovedBrowser(): String =
        policy.approvedBrowser?.takeIf { it.isNotBlank() } ?: DEFAULT_BROWSER

    /** True when non-approved browsers should be locked. Follows the site
     *  admin's switch alone — safe browsing used to imply it, so turning the
     *  switch off changed nothing and every other browser stayed blocked. */
    fun shouldLockOtherBrowsers(): Boolean = policy.blockOtherBrowsers

    const val DEFAULT_BROWSER = "com.android.chrome"

    /** Alias domains that are really the same service, so blocking the canonical
     *  domain also blocks the alias (e.g. blocking youtube.com blocks youtu.be). */
    private val DOMAIN_ALIASES = mapOf(
        "youtu.be" to "youtube.com",
    )

    /** YouTube website domains. YouTube is always blocked in the *browser* (the
     *  child should use the YouTube app instead) — this is built in and does NOT
     *  depend on the parent's blocklist or the safe-browsing switch. It only
     *  affects the browser: the YouTube app is unaffected because app traffic
     *  never goes through this address-bar check. */
    private val YOUTUBE_WEB_DOMAINS = setOf(
        "youtube.com",
        "youtu.be",
        "youtube-nocookie.com",
        "youtubekids.com",
    )

    /** True if [host] is a YouTube website domain (always blocked in browser). */
    fun isYoutubeWeb(host: String): Boolean {
        val d = host.lowercase().removePrefix("www.").removePrefix("m.")
        return YOUTUBE_WEB_DOMAINS.any { d == it || d.endsWith(".$it") }
    }

    /** Search-engine result pages. These merely *list links* (with snippets that
     *  may contain a keyword), so the page-text content scan is skipped for them
     *  to avoid blocking the results page itself — only the site the child
     *  actually opens is content-scanned. URL/host blocking still applies. */
    private val SEARCH_ENGINES = setOf(
        "bing.com", "duckduckgo.com", "search.yahoo.com", "ecosia.org",
        "startpage.com", "qwant.com", "baidu.com", "ask.com",
        "search.brave.com", "searx.be", "swisscows.com", "mojeek.com",
    )

    /** True if [host] is a search-engine results page (Google incl. all TLDs,
     *  Yandex, or one of [SEARCH_ENGINES]). */
    fun isSearchEngine(host: String): Boolean {
        val h = host.lowercase().removePrefix("www.")
        if (h == "google.com" || h.startsWith("google.") || h.contains(".google.")) {
            return true
        }
        if (h == "bing.com" || h.startsWith("yandex.")) return true
        return SEARCH_ENGINES.any { h == it || h.endsWith(".$it") }
    }

    /** True if [host] should be blocked (exact/subdomain site or a category hit). */
    fun isBlocked(host: String): Boolean = reasonFor(host) != null

    /**
     * Why [host] is blocked — a category id, [REASON_BLOCKLIST] for the parent's
     * own list, or null when it's allowed. The parent app turns this into a tag.
     */
    fun reasonFor(host: String): String? {
        val p = policy // one snapshot for the whole decision
        if (!p.enabled) return null
        val d = host.lowercase().removePrefix("www.")
        if (d.isEmpty()) return null
        if (matchesBlockedSite(d, p.blockedSites)) return REASON_BLOCKLIST
        // Treat a known alias as its canonical domain (youtu.be -> youtube.com).
        val canonical = DOMAIN_ALIASES[d]
        if (canonical != null && matchesBlockedSite(canonical, p.blockedSites)) {
            return REASON_BLOCKLIST
        }
        return CategoryFeed.categoryOf(d, p.enabledCategories)
    }

    const val REASON_BLOCKLIST = "blocklist"
    const val REASON_YOUTUBE = "youtube"
    const val REASON_KEYWORD = "keyword"
    const val REASON_CONTENT = "content"

    /** Curated per-category seed domains — blocked instantly and offline, before
     *  the live category feeds download. Shared with [WebFilterVpnService]. */
    val CATEGORY_SEEDS: Map<String, List<String>> = mapOf(
        "adult" to listOf(
            "pornhub.com", "xvideos.com", "xnxx.com", "xhamster.com",
            "redtube.com", "youporn.com",
        ),
        "gambling" to listOf(
            "bet365.com", "pokerstars.com", "williamhill.com", "888casino.com",
        ),
        "social" to listOf(
            "facebook.com", "instagram.com", "twitter.com", "x.com",
            "tiktok.com", "snapchat.com",
        ),
        "drugs" to listOf("leafly.com", "weedmaps.com"),
        "weapons" to listOf("armslist.com"),
        "violence" to listOf("liveleak.com"),
    )

    /** The seed category [domain] falls in (any category), or null. */
    fun seedCategoryOf(domain: String): String? {
        val d = domain.lowercase().removePrefix("www.")
        for ((cat, seeds) in CATEGORY_SEEDS) {
            if (seeds.any { d == it || d.endsWith(".$it") }) return cat
        }
        return null
    }

    /**
     * Best-effort category for tagging a block in the parent's activity list.
     * Checks every source regardless of which categories are enabled — a
     * gambling site the parent listed manually should still read "Gambling".
     * Never used for the blocking decision itself.
     */
    fun categoryHint(host: String): String? {
        val d = host.lowercase().removePrefix("www.")
        if (d.isEmpty()) return null
        return seedCategoryOf(d)
            ?: ContentFilter.matchHost(d)?.let { ContentFilter.categoryOf(it) }
            ?: CategoryFeed.categoryOfAny(d)
    }

    /** True if [domain] is on the blocklist, as an exact or parent domain. */
    private fun matchesBlockedSite(domain: String, sites: Set<String>): Boolean =
        sites.any { domain == it || domain.endsWith(".$it") }
}

File: child_windows/src/GuardNest.Core/Web/WebFilter.cs
namespace GuardNest.Core.Web;

/// <summary>Why a page was refused, mirroring the reasons the parent app labels.</summary>
public static class BlockReason
{
    public const string Blocklist = "blocklist";
    public const string Keyword = "keyword";
    public const string Youtube = "youtube";
}

/// <summary>One consistent view of the filter, replaced wholesale rather than mutated.</summary>
public sealed record WebPolicy(
    IReadOnlySet<string> BlockedSites,
    IReadOnlySet<string> BlockedCategories,
    bool BlockOtherBrowsers,
    bool AllowIncognito,
    string ApprovedBrowser,
    bool BlockYoutubeWeb)
{
    /// <summary>
    /// YouTube in a browser is blocked on Android so the child uses the app,
    /// where the parent can see what was watched. Windows has no such app, so
    /// the same rule would ban YouTube outright instead of redirecting it; the
    /// default here is to allow it and capture the watch history from the
    /// window title. A site admin can still switch the ban on.
    /// </summary>
    public static readonly WebPolicy Default = new(
        new HashSet<string>(StringComparer.OrdinalIgnoreCase),
        WebFilter.ProtectiveCategories,
        BlockOtherBrowsers: false,
        AllowIncognito: false,
        ApprovedBrowser: "msedge.exe",
        BlockYoutubeWeb: false);
}

/// <summary>
/// The blocking decision, ported from WebFilter.kt. Reading one immutable
/// snapshot per decision means a policy update can never be observed half
/// applied — new categories with the old site list, say.
/// </summary>
public sealed class WebFilter
{
    /// Blocked for everyone, always; not even the site admin can turn these off.
    public static readonly IReadOnlySet<string> ProtectiveCategories =
        new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "adult", "gambling", "drugs", "weapons", "violence",
        };

    /// Blocking the canonical domain has to block the alias too.
    private static readonly Dictionary<string, string> Aliases = new(StringComparer.OrdinalIgnoreCase)
    {
        ["youtu.be"] = "youtube.com",
    };

    private static readonly HashSet<string> YoutubeWeb = new(StringComparer.OrdinalIgnoreCase)
    {
        "youtube.com", "youtu.be", "youtube-nocookie.com", "youtubekids.com",
    };

    private static readonly HashSet<string> SearchEngines = new(StringComparer.OrdinalIgnoreCase)
    {
        "bing.com", "duckduckgo.com", "search.yahoo.com", "ecosia.org",
        "startpage.com", "qwant.com", "baidu.com", "ask.com",
        "search.brave.com", "searx.be", "swisscows.com", "mojeek.com",
    };

    private readonly CategoryFeed _categories;
    private volatile WebPolicy _policy = WebPolicy.Default;
    private volatile IReadOnlySet<string> _keywords =
        new HashSet<string>(StringComparer.OrdinalIgnoreCase);

    public WebFilter(CategoryFeed categories)
    {
        _categories = categories;
    }

    public WebPolicy Policy => _policy;

    public void Apply(WebPolicy policy) => _policy = policy;

    public void ApplyKeywords(IReadOnlySet<string> keywords) => _keywords = keywords;

    /// <summary>True for a results page, whose link snippets must not be keyword-matched.</summary>
    public static bool IsSearchEngine(string host)
    {
        var value = Domains.Normalize(host);
        if (value == "google.com" || value.StartsWith("google.") || value.Contains(".google.")) return true;
        if (value.StartsWith("yandex.")) return true;
        return SearchEngines.Any(engine => Domains.Covers(engine, value));
    }

    public static bool IsYoutube(string host)
    {
        var value = Domains.Normalize(host);
        return YoutubeWeb.Any(domain => Domains.Covers(domain, value));
    }

    public string? ReasonFor(string address) => ReasonFor(address, "");

    /// <summary>
    /// Why this address is refused, or null when it is allowed. Keywords are
    /// looked for in the page itself, not just the address: the words that make
    /// a page unsafe are usually in its text and its title and nowhere in the
    /// URL. Reading the rendered page is also what makes this work over HTTPS.
    /// </summary>
    public string? ReasonFor(string address, string pageTitle, string pageText = "")
    {
        var host = Domains.HostOf(address);
        if (host.Length == 0) return null;

        // Windows has no separate YouTube app to redirect into. Always allow
        // browser YouTube and capture its searches/watch history instead.
        if (IsYoutube(host)) return null;

        var policy = _policy;
        if (Domains.MatchesAny(host, policy.BlockedSites)) return BlockReason.Blocklist;
        if (Aliases.TryGetValue(host, out var canonical) &&
            Domains.MatchesAny(canonical, policy.BlockedSites))
        {
            return BlockReason.Blocklist;
        }

        var category = _categories.CategoryOf(host, policy.BlockedCategories);
        if (category is not null) return category;

        // A results page lists other people's words; matching keywords there
        // would block the search itself rather than anything the child opened.
        if (IsSearchEngine(host)) return null;
        return KeywordIn(Unescape(address)) ?? KeywordIn(pageTitle) ?? KeywordIn(pageText);
    }

    private string? KeywordIn(string content)
    {
        var keywords = _keywords;
        if (keywords.Count == 0 || content.Length == 0) return null;
        var text = content.ToLowerInvariant();
        return keywords.Any(word => ContainsWord(text, word)) ? BlockReason.Keyword : null;
    }

    private static string Unescape(string address)
    {
        // A percent sign that is not a valid escape is ordinary text here.
        try { return Uri.UnescapeDataString(address); }
        catch (UriFormatException) { return address; }
    }

    /// <summary>
    /// Whole-word match only. A plain substring test blocks essex.gov.uk for
    /// "sex", and a parental control that blocks a council website is one the
    /// family switches off.
    /// </summary>
    internal static bool ContainsWord(string text, string word)
    {
        if (word.Length == 0) return false;
        var index = text.IndexOf(word, StringComparison.Ordinal);
        while (index >= 0)
        {
            var beforeOk = index == 0 || !char.IsLetterOrDigit(text[index - 1]);
            var end = index + word.Length;
            var afterOk = end >= text.Length || !char.IsLetterOrDigit(text[end]);
            if (beforeOk && afterOk) return true;
            index = text.IndexOf(word, index + 1, StringComparison.Ordinal);
        }
        return false;
    }

    /// <summary>A label for the parent's activity list, even for a manually listed site.</summary>
    public string? CategoryHint(string host) =>
        _categories.CategoryOfAny(Domains.Normalize(host));
}
